salt/_pillar/azureblob.py

```diff
@@ -0,0 +1,449 @@
+    if not os.path.exists(os.path.dirname(file_path)):
+        os.makedirs(os.path.dirname(file_path))
+
+    return file_path
+
+
+def _get_containers_cache_filename(container, prefix):
+    '''
+    .. versionadded:: Sodium
+
+    Return the filename of the cache for container contents. Create the path if it does not exist.
+
+    :param container: The name of the target Azure Blob Container.
+
+    '''
+    cache_dir = _get_cache_dir()
+    if not os.path.exists(cache_dir):
+        os.makedirs(cache_dir)
+
+    return os.path.join(cache_dir, '{0}-{1}-files.cache'.format(container, prefix))
+
+
+def _refresh_containers_cache_file(connection_string, container, cache_file,
+                                   multiple_env=False, environment='base'):
+    '''
+    .. versionadded:: Sodium
+
+    Downloads the entire contents of an Azure storage container to the local filesystem.
+
+    :param connection_string: The connection string to use to access the specified Azure Blob Container.
+
+    :param container: The name of the target Azure Blob Container.
+
+    :param cache_file: The path of where the file will be cached.
+
+    :param multiple_env: Specifies whether the pillar should interpret top level folders as pillar environments.
+
+    :param environment: Specifies which environment the container represents when in single environment mode. This is
+        ignored if multiple_env is set as True.
+
+    '''
+    try:
+        # Create the BlobServiceClient object which will be used to create a container client
+        blob_service_client = BlobServiceClient.from_connection_string(connection_string)
+
+        # Create the ContainerClient object
+        container_client = blob_service_client.get_container_client(container)
+    except Exception as exc:
+        log.error('Exception: %s', exc)
+        return False
+
+    metadata = {}
+
+    def _walk_blobs(saltenv='base', prefix=None):
+        # Walk the blobs in the container with a generator
+        blob_list = container_client.walk_blobs(name_starts_with=prefix)
+
+        # Iterate over the generator
+        while True:
+            try:
+                blob = next(blob_list)
+            except StopIteration:
+                break
+
+            log.debug('Raw blob attributes: %s', blob)
+
+            # Directories end with "/". Make them locally.
+            if blob.name.endswith('/'):
+                # Recurse into the directory
+                _walk_blobs(prefix=blob.name)
+
+                continue
+
+            if multiple_env:
+                saltenv = ('base' if prefix == '.' else prefix)
+
+            if saltenv not in metadata:
+                metadata[saltenv] = {}
+
+            if container not in metadata[saltenv]:
+                metadata[saltenv][container] = []
+
+            metadata[saltenv][container].append(blob)
+
+    _walk_blobs(saltenv=environment)
+
+    # write the metadata to disk
+    if os.path.isfile(cache_file):
```

Comment:
Does s3 use the `os` library for this, or salt util/exec modules?

Reply:
The s3 ext_pillar use the os library for that. Those lines are actually copied directly from the s3 ext_pillar.